### PR TITLE
feat(tiering): Registered buffers

### PR DESCRIPTION
**Pull request by dranikpg**
* Use UringBuf to optionally issue prepared read/write
* Remove IoMgr functions no longer in use

### CODE REVIEW

File: src/server/tiering/disk_storage.cc [hunk 1]
```diff
@@ -8,12 +8,55 @@
...
+
+}  // anonymous namespace
+
 std::error_code DiskStorage::Open(std::string_view path) {
   RETURN_ON_ERR(io_mgr_.Open(path));
   alloc_.AddStorage(0, io_mgr_.Span());
+
+  auto* up = static_cast<UringProactor*>(ProactorBase::me());
+  up->RegisterBuffers(512_KB);
+
```
Comment (dranikpg) [inline src/server/tiering/disk_storage.cc:59]: Who's in charge of doing this / deciding on the size?

File: src/server/tiering/test_common.h [hunk 2]
```diff
@@ -15,22 +15,17 @@ namespace dfly::tiering {
...
+  }
 
-  static std::unique_ptr<util::ProactorPool> pp_;
-};
-
-std::unique_ptr<util::ProactorPool> PoolTestBase::pp_ = nullptr;
+  void TearDown() override {
+    pp_->Stop();
+    pp_.reset();
+  }
```
Comment (dranikpg) [inline src/server/tiering/test_common.h:27]: I've shot myself in the foot like you predicted in the helio PR 😆

File: src/server/tiered_storage.cc [hunk 3]
```diff
@@ -68,17 +81,14 @@ class TieredStorageV2::ShardOpManager : public tiering::OpManager {
...
-  void ReportStashed(EntryId id, tiering::DiskSegment segment) override {
-    if (holds_alternative<string_view>(id)) {
-      SetExternal(get<string_view>(id), segment);
-    } else {
-      for (const auto& [sub_key, sub_segment] :
-           ts_->bins_->ReportStashed(get<tiering::SmallBins::BinId>(id), segment))
-        SetExternal(string_view{sub_key}, sub_segment);
-    }
+  void ReportStashed(EntryId id, tiering::DiskSegment segment, error_code ec) override {
+    if (ec)
```
Comment (romange) [inline src/server/tiered_storage.cc:85]: we should report the io errors

File: src/server/tiering/disk_storage.cc [hunk 4]
```diff
@@ -8,12 +8,55 @@
...
+}
+
+}  // anonymous namespace
+
 std::error_code DiskStorage::Open(std::string_view path) {
   RETURN_ON_ERR(io_mgr_.Open(path));
   alloc_.AddStorage(0, io_mgr_.Span());
+
+  auto* up = static_cast<UringProactor*>(ProactorBase::me());
+  up->RegisterBuffers(512_KB);
```
Comment (romange) [inline src/server/tiering/disk_storage.cc:58]: check errors

File: src/server/tiering/disk_storage.cc (hunk 1 above)
Comment (romange) [reply src/server/tiering/disk_storage.cc:59]: either constexpr or a flag with a sane default

File: src/server/tiered_storage.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/tiered_storage.cc:85]: you mean add them to stats? log them?

Comment (romange) [reply src/server/tiered_storage.cc:85]: we can start with logging them

File: src/server/tiering/disk_storage.cc [hunk 5]
```diff
@@ -4,16 +4,68 @@
...
+
+UringBuf PrepareBuf(size_t size) {
+  DCHECK_EQ(ProactorBase::me()->GetKind(), ProactorBase::IOURING);
+  auto* up = static_cast<UringProactor*>(ProactorBase::me());
+
+  UringBuf buf;
+  if (auto borrowed = up->RequestBuffer(size); borrowed)
+    return *borrowed;
+  else
+    return AllocateTmpBuf(size);
```
Comment (romange) [inline src/server/tiering/disk_storage.cc:44]: I do not mind whatever policy you implement but we should track this so we could learn from this

File: src/server/tiering/disk_storage.h [hunk 6]
```diff
@@ -36,6 +36,7 @@ class DiskStorage {
   // Request bytes to be stored, cb will be called with assigned segment on completion. Can block to
   // grow backing file. Returns error code if operation failed  immediately (most likely it failed
   // to grow the backing file) or passes an empty segment if the final write operation failed.
+  // Bytes must not outlive the object call and can be dropped before cb is resolved
```
Comment (romange) [inline src/server/tiering/disk_storage.h:39]: `Bytes must not outlive` sounds like they have to be deleted before. Maybe just keep the second part of the sentence.

**Review by romange — APPROVED** [review]

File: src/server/tiered_storage.cc [hunk 7]
```diff
@@ -167,15 +179,18 @@ void TieredStorageV2::Stash(string_view key, PrimeValue* value) {
...
   } else if (auto bin = bins_->Stash(key, value_sv); bin) {
-    if (auto ec = op_manager_->Stash(bin->first, bin->second); ec) {
-      for (const string& key : bins_->ReportStashAborted(bin->first))
-        op_manager_->ClearIoPending(key);  // clear IO_PENDING flag
-    }
+    id = bin->first;
+    ec = op_manager_->Stash(bin->first, bin->second);
   }
+
+  if (ec)
```
Comment (romange) [inline src/server/tiered_storage.cc:192]: I think you are missing error logs here as well.

File: src/server/tiering/disk_storage.cc (hunk 5 above)
Comment (romange) [reply src/server/tiering/disk_storage.cc:46]: ping

Comment (dranikpg) [reply src/server/tiering/disk_storage.cc:46]: AllocateTmpBuf has a vlog

**Review by romange — APPROVED** [review]

**Pull request merged**